j5/vision/markers.py

```diff
@@ -0,0 +1,203 @@
+        """Provides the coordiante in the form of a cylindrical coordinate."""
+        return self
+
+    def as_spherical(self) -> Spherical:
+        """Provides the coordiante in the form of a spherical coordinate."""
+        return Spherical(np.sqrt(self.p**2 + self.z**2),
+                         np.arctan2(self.p, self.z),
+                         self.phi)
+
+
+class Spherical(NamedTuple):
+    """A Spherical Coordinate.
+
+    r     := radial distance
+    theta := polar angle (radians)
+    phi   := azimuth angle (radians)
+    """
+
+    r: float
+    theta: float
+    phi: float
+
+    def as_cartesian(self) -> Cartesian:
+        """Provides the coordiante in the form of a cartesian coordinate."""
+        return Cartesian(self.r * np.cos(self.phi) * np.sin(self.theta),
+                         self.r * np.sin(self.phi) * np.sin(self.theta),
+                         self.r * np.cos(self.theta))
+
+    def as_cylindrical(self) -> Cylindrical:
+        """Provides the coordiante in the form of a cylindrical coordinate."""
+        return Cylindrical(self.r * np.sin(self.theta),
+                           self.phi,
+                           self.r * np.cos(self.theta))
+
+    def as_spherical(self) -> Spherical:
+        """Provides the coordiante in the form of a spherical coordinate."""
+        return self
+
+
+Coordinate = Union[Cylindrical, Spherical, Cartesian]
+
+
+class Markers():
+    """Markers Class.
+
+    The position is stored in self.__position.
+    This is a multi-dementional numpy array.
+    The first dimesion is the marker index.
+    The second dimension is the axis.
+    Example:
+        [ [x1, y1, z1],
+          [x2, y2, z2],
+          [x3, y3, z3],
+          [x4, y4, z4] ]
+    """
+
+    def __init__(self):
+        self.__positions = None
+
+    @property
+    def positions(self) -> List[Cartesian]:
+        """Returns the makers positions."""
+        return [Cartesian(*coordinate) for coordinate in self.__positions]
+
+    def add_markers_cartesian(self, given_positions: List[Cartesian]) -> None:
+        """Add marker(s) by providing their cartesian coordinates."""
+        given_positions = np.array(given_positions)
+        self.__append_positions(given_positions)
+
+    def add_markers(self, given_positions: List[Coordinate]) -> None:
+        """Add marker(s) by providing their coordinates."""
+        given_positions = np.array([position.to_cartesian()
+                                    for position in given_positions
+                                    ])
+        self.__append_positions(given_positions)
+
+    def __append_positions(self, given_positions):
+        """Appends a numpy array of markers to the current array (if one exists)."""
+        if self.__positions is not None:
+            self.__positions = np.concatenate([self.__positions, given_positions])
+        else:
+            self.__positions = given_positions
+
+    def translate_space(self, delta_x, delta_y, delta_z):
```

Comment:
Why are we supporting mutation of the markers? Markers represent a position in physical space, why can't be mutated by the robot in the way these methods imply.

Reply:
My plan was to keep the robot at the origin. The mutations allow the robot to guess where the markers will be after a movement without having to wait for the vision system to update marker locations. This means robots speed isn't limited by the vision system.

Reply:
Whats the verdict? Are space transforms wanted?

Reply:
I personally don't like the markers being mutable. I think we can implement a function that maps over a MarkerList to return a new list quite easily also.